Modules/Core/Common/include/itkImageScanlineConstIterator.h

```diff
@@ -260,6 +260,12 @@ class ITK_TEMPLATE_EXPORT ImageScanlineConstIterator : public ImageConstIterator
   OffsetValueType m_SpanBeginOffset{}; // one pixel the beginning of the scanline
   OffsetValueType m_SpanEndOffset{};   // one pixel past the end of the scanline
 };
+
+// Deduction guide for class template argument deduction (CTAD).
+template <typename TImage>
+ImageScanlineConstIterator(SmartPointer<const TImage>, const typename TImage::RegionType &)
```

Comment:
This type of deduction guide is needed for most classes and templated methods in ITK, in order to avoid this less comprehensible code:
https://github.com/InsightSoftwareConsortium/ITK/blob/v5.3.0/Modules/IO/ImageBase/include/itkImageFileWriter.h#L245-L254

Reply:
Before we had that implementation with R-value references, I used to have two signatures of `WriteImage`, one with SmartPointer and one with RawPointer parameter (and one signature delegate to the other). I thought that CTAD could be used to write more elegant and comprehensible code which does not need to resort to R-value references.

Reply:
Apparently, there is a paper that proposes deduction guides for function templates: Mike Spertus, 2018-10-08, 
[Improving function templates with Class Template Argument Deduction](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p1167r0.html)  Is that what you mean? Not yet included with C++17, obviously.

Reply:
Of course I'm still in favor of pass-by-reference, instead of pass-by-pointer, when the passed object must be valid (not "null") and the ownership is not "passed" to the function. (That is, when the lifetime of the object is not _extended_ by passing it to the function.)

Reply:
I didn't know that proposal existed, but I want it 😄

For now, we could add CTAD guides for classes, where useful.